pep/002-mixin-versioning.md

```diff
@@ -0,0 +1,355 @@
+---
+title: "Mixin Versioning"
+number: "NNN"
+status: "provisional"
+owner: "@carolynvs"
+authors:
+  - "@carolynvs"
+  - "@sestegra"
+created: "2021-02-21"
+type: "feature"
+---
+
+## Abstract
+
+Bundles should be able to specify not only the mixin used, but its version and source.
+Porter should use that information to automatically download the mixins declared by the bundle.
+
+
+## Motivation
+
+Currently bundle authors manage the mixins used within a bundle by hand. 
+They must manually install each mixin using homegrown scripts. 
+Unfortunately the mixin cache is stored in PORTER_HOME, it is impossible to work on bundles that use different versions of the same mixin.
+Porter really needs to manage the mixin cache and appropriately communicating with the correct mixin version targeted by the bundle.
+
+* Mixin authors need to adjust their mixin declaration schema to allow for declaring the mixin version and release a new version of their mixin(s) supporting this proposal.
+* Bundle authors should start defining the mixin version when they declare mixins in porter.yaml. If they have extra scripts that they were running before `porter build` to ensure the right versions are in the global mixin cache, that is no longer necessary.
+* End users will now be able to see the exact version of the mixins used by a bundle.
+
+
+## Rationale
+
+<!--
+The rationale fleshes out the specification by describing why particular design
+decisions were made. It should describe alternate designs that were considered
+and related work.
+
+The rationale should provide evidence of consensus within the community and
+discuss important objections or concerns raised during discussion.
+-->
+A lot of this design is driven by how mature tools in the ecosystem address this problem.
+Since Porter's architecture is similar to Terraform, specifically Porter mixins and Terraform providers/modules, I have used Terraform's solutions to caching and locking the mixins used in a bundle to understand gaps in our own solution.
+
+This feature was originally suggested at https://github.com/getporter/porter/issues/975 and by @sestegra in https://github.com/getporter/porter/discussions/1311.
+
+
+## Specification
+
+### Mixin Cache
+
+First we need to alter the mixin cache structure in PORTER_HOME/mixins.
+The new file structure allows for caching multiple versions of a mixin concurrently, and allows for disambiguation when the same mixin version is cached from different sources.
+For example, two users can publish mixins with the same name and version and Porter should be able to distinguish between the two.
+
+```
+PORTER_HOME/
+  mixins/
+    SOURCE/ # May be nested directories, e.g. cdn.porter.sh/mixins/exec
+      NAME/
+        VERSION/
+          NAME # client binary
+          runtimes/
+            NAME-GOOS-GOARCH # Runtime binary, e.g. aws-linux-amd64
+```
+
```

Comment:
should we consider this behavior of dynamic downloading that was discussed here?  https://github.com/getporter/kubernetes-mixin/issues/18

Reply:
I am not sure how these are related?

That issue is about the kubernetes mixin or perhaps extending the design to a general pattern for mixins that download a tool's CLI at runtime, ensuring that the client version matches the cluster's.

This section is specifying the storage format on client filesystems, i.e. when I install the binary for a mixin (not a tool's CLI), where are the files saved. True, Porter is downloading the binary just in time (dynamically) so while that issue may appear related, the conditions for determining which version to download and from which source are very different I think.